scipy/stats/_stats_py.py

```diff
@@ -3908,34 +3922,88 @@ def f_oneway(*samples, axis=0):
     all_same_const = (_first(alldata, axis) == alldata).all(axis=axis)
 
-    # Determine the mean of the data, and subtract that from all inputs to a
-    # variance (via sum_of_sq / sq_of_sum) calculation.  Variance is invariant
-    # to a shift in location, and centering all data around zero vastly
-    # improves numerical stability.
-    offset = alldata.mean(axis=axis, keepdims=True)
-    alldata = alldata - offset
+    if not isinstance(equal_var, bool):
+        raise TypeError("Expected a boolean value for 'equal_var'")
 
-    normalized_ss = _square_of_sums(alldata, axis=axis) / bign
+    if equal_var:
+        # Determine the mean of the data, and subtract that from all inputs to a
+        # variance (via sum_of_sq / sq_of_sum) calculation.  Variance is invariant
+        # to a shift in location, and centering all data around zero vastly
+        # improves numerical stability.
+        offset = alldata.mean(axis=axis, keepdims=True)
+        alldata = alldata - offset
+
+        normalized_ss = _square_of_sums(alldata, axis=axis) / bign
+
+        sstot = _sum_of_squares(alldata, axis=axis) - normalized_ss
+
+        ssbn = 0
+        for sample in samples:
+            smo_ss = _square_of_sums(sample - offset, axis=axis)
+            ssbn = ssbn + smo_ss / sample.shape[axis]
+
+        # Naming: variables ending in bn/b are for "between treatments", wn/w are
+        # for "within treatments"
+        ssbn = ssbn - normalized_ss
+        sswn = sstot - ssbn
+        dfbn = num_groups - 1
+        dfwn = bign - num_groups
+        msb = ssbn / dfbn
+        msw = sswn / dfwn
+        with np.errstate(divide='ignore', invalid='ignore'):
+            f = msb / msw
 
-    sstot = _sum_of_squares(alldata, axis=axis) - normalized_ss
+        prob = special.fdtrc(dfbn, dfwn, f)   # equivalent to stats.f.sf
 
-    ssbn = 0
-    for sample in samples:
-        smo_ss = _square_of_sums(sample - offset, axis=axis)
-        ssbn = ssbn + smo_ss / sample.shape[axis]
-
-    # Naming: variables ending in bn/b are for "between treatments", wn/w are
-    # for "within treatments"
-    ssbn = ssbn - normalized_ss
-    sswn = sstot - ssbn
-    dfbn = num_groups - 1
-    dfwn = bign - num_groups
-    msb = ssbn / dfbn
-    msw = sswn / dfwn
-    with np.errstate(divide='ignore', invalid='ignore'):
-        f = msb / msw
+    # if population variances are not equeal, perform Welch ANOVA
+    # the related issue: https://github.com/scipy/scipy/issues/11122
+    else:
+        # calculate basic statistics for each sample
+        samples_arrs = np.array(samples)
+
+        # adjust the shape of the data based on the arrary dimension for the calculation
+        # along with input axis
+        if samples_arrs.ndim <= 2:
+            samples_arrs = samples_arrs.T
+        else:
+            samples_arrs = samples_arrs.transpose(
+                *range(1, samples_arrs.ndim - 1), 0, -1
+            )
+
+        means = samples_arrs.mean(axis=axis)
```

Comment:
Just do something like:
```python3
means = [np.mean(sample, axis=axis) for sample in samples]
```
Same below; please just use list comprehensions instead of `apply_along_axis`.

After the reduction (of $n$ observations to a single number), you can put the results in an array if convenient.